File: agent_project/activity.py
"""Shared activity-event contract.

A single normalized event envelope (`type="activity"`) describes every
unit of agent work — chat tool calls, research tool calls, workflow
substeps, etc. — so the frontend has one model to render.

Legacy events (`tool_call_start/end/error`, `workflow_step`, ...) still
ship alongside activity events during the migration window. Once parity
is confirmed across all tools/workflows the legacy emissions can be
deleted.
"""

from __future__ import annotations

from typing import Any, Literal, TypedDict

ACTIVITY_EVENT_TYPE = "activity"

# Coarse buckets — drives icon / colour / grouping in the UI.
ActivityKind = Literal[
    "tool",            # individual tool invocation (search_web, calculator, ...)
    "workflow",        # parent workflow span (e.g. DCF root)
    "workflow_step",   # substep inside a workflow
    "node",            # graph node-level activity (plan, synthesize, ...)
]

# Lifecycle states a single activity goes through.
ActivityStatus = Literal[
    "started",
    "running",      # alias for "in progress"; prefer "started" + "completed"
    "completed",
    "skipped",
    "error",
    "awaiting_input",
]

# Where the activity originated, so the UI can scope it correctly.
ActivityScope = Literal[
    "chat",            # inside a chat-mode ReAct turn
    "research",        # inside the research subgraph
    "workflow",        # inside a domain workflow (DCF, future: comps/LBO)
]


class ActivityEvent(TypedDict, total=False):
    """Canonical activity envelope sent over SSE.

    All keys except those marked Required are optional. Producers should
    use the `make_activity` helper rather than constructing this dict by
    hand to avoid drift.
    """
    type: str                 # always ACTIVITY_EVENT_TYPE
    activity_id: str          # stable id for matching start->end
    parent_activity_id: str   # nests workflow steps under their workflow
    kind: ActivityKind        # bucket for icon / grouping
    name: str                 # raw identifier (tool name, step id, ...)
    display_label: str        # optional override; UI may still re-label
    scope: ActivityScope      # research | chat | workflow
    status: ActivityStatus    # lifecycle state
    step_id: str              # research step id, or "chat" for chat scope
    started_at: float         # epoch seconds (set on the *first* emit)
    ended_at: float           # epoch seconds (set on terminal emit)
    summary: str              # one-line human readable
    args_preview: str         # short string preview of inputs
    confidence_label: str     # workflow-only; surfaces trust signal
    flag_count: int           # workflow-only; total quality flags
    error: str                # populated when status == "error"
    meta: dict[str, Any]      # free-form payload (links, ids, etc.)


def make_activity(
    *,
    activity_id: str,
    kind: ActivityKind,
    name: str,
    scope: ActivityScope,
    status: ActivityStatus,
    parent_activity_id: str | None = None,
    display_label: str | None = None,
    step_id: str | None = None,
    started_at: float | None = None,
    ended_at: float | None = None,
    summary: str | None = None,
    args_preview: str | None = None,
    confidence_label: str | None = None,
    flag_count: int | None = None,
    error: str | None = None,
    meta: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Build an ActivityEvent payload, omitting unset fields."""
    payload: dict[str, Any] = {
        "type": ACTIVITY_EVENT_TYPE,
        "activity_id": activity_id,
        "kind": kind,
        "name": name,
        "scope": scope,
        "status": status,
    }
    if parent_activity_id:
        payload["parent_activity_id"] = parent_activity_id
    if display_label:
        payload["display_label"] = display_label
    if step_id:
        payload["step_id"] = step_id
    if started_at is not None:
        payload["started_at"] = started_at
    if ended_at is not None:
        payload["ended_at"] = ended_at
    if summary:
        payload["summary"] = summary
    if args_preview:
        payload["args_preview"] = args_preview
    if confidence_label:
        payload["confidence_label"] = confidence_label
    if flag_count is not None:
        payload["flag_count"] = flag_count
    if error:
        payload["error"] = error
    if meta:
        payload["meta"] = meta
    return payload

File: agent_project/frontend/src/lib/activity.ts
// Shared ActivityEvent contract — mirror of agent_project/activity.py.
// Keep these two files in lockstep.

export const ACTIVITY_EVENT_TYPE = 'activity'

export type ActivityKind =
  | 'tool'
  | 'workflow'
  | 'workflow_step'
  | 'node'

export type ActivityStatus =
  | 'started'
  | 'running'
  | 'completed'
  | 'skipped'
  | 'error'
  | 'awaiting_input'

export type ActivityScope = 'chat' | 'research' | 'workflow'

export interface ActivityEvent {
  type: typeof ACTIVITY_EVENT_TYPE
  activity_id: string
  kind: ActivityKind
  name: string
  scope: ActivityScope
  status: ActivityStatus
  parent_activity_id?: string
  display_label?: string
  step_id?: string
  started_at?: number
  ended_at?: number
  summary?: string
  args_preview?: string
  confidence_label?: string
  flag_count?: number
  error?: string
  meta?: Record<string, unknown>
}

// Frontend-side aggregation: a single Activity can receive multiple
// events (start -> running -> completed). The store collapses them into
// one entry keyed by activity_id, preserving the most recent status.
export interface ActivityEntry {
  activity_id: string
  parent_activity_id?: string
  kind: ActivityKind
  name: string
  display_label?: string
  scope: ActivityScope
  status: ActivityStatus
  step_id?: string
  started_at?: number
  ended_at?: number
  summary: string
  args_preview: string
  confidence_label?: string
  flag_count?: number
  error?: string
  meta?: Record<string, unknown>
}

export function isActivityEvent(data: Record<string, unknown>): boolean {
  return data?.type === ACTIVITY_EVENT_TYPE && typeof data.activity_id === 'string'
}

/** Map an ActivityEntry's lifecycle status into the legacy `ToolCall.status`
 *  triplet (`running` / `done` / `error`) used by `StepCard`,
 *  `ResearchStepsTrace`, and the activity row renderer. */
export function activityStatusToToolStatus(
  status: ActivityStatus,
): 'running' | 'done' | 'error' {
  if (status === 'completed' || status === 'skipped') return 'done'
  if (status === 'error') return 'error'
  return 'running'
}

/**
 * Merge an incoming ActivityEvent into an existing entry list.
 *
 * - If `activity_id` already exists, fields from the new event win
 *   (status, ended_at, summary, error, ...) but timestamps and other
 *   already-set fields on the existing entry are preserved when the new
 *   event omits them.
 * - Otherwise the entry is appended.
 *
 * Returned list is a new array (immutable update friendly).
 */
export function mergeActivity(
  entries: ActivityEntry[],
  event: ActivityEvent,
): ActivityEntry[] {
  const idx = entries.findIndex(e => e.activity_id === event.activity_id)
  const incoming: ActivityEntry = {
    activity_id: event.activity_id,
    parent_activity_id: event.parent_activity_id,
    kind: event.kind,
    name: event.name,
    display_label: event.display_label,
    scope: event.scope,
    status: event.status,
    step_id: event.step_id,
    started_at: event.started_at,
    ended_at: event.ended_at,
    summary: event.summary ?? '',
    args_preview: event.args_preview ?? '',
    confidence_label: event.confidence_label,
    flag_count: event.flag_count,
    error: event.error,
    meta: event.meta,
  }

  if (idx < 0) return [...entries, incoming]

  const prev = entries[idx]
  const merged: ActivityEntry = {
    ...prev,
    ...incoming,
    started_at: prev.started_at ?? incoming.started_at,
    summary: incoming.summary || prev.summary,
    args_preview: incoming.args_preview || prev.args_preview,
    display_label: incoming.display_label ?? prev.display_label,
    confidence_label: incoming.confidence_label ?? prev.confidence_label,
    flag_count: incoming.flag_count ?? prev.flag_count,
    error: incoming.error ?? prev.error,
    meta: incoming.meta ?? prev.meta,
  }
  const next = [...entries]
  next[idx] = merged
  return next
}
